Rewrite Gutenberg configuration file from Ruby to YAML

mokagio opened the pull request:
This should prove easier to manipulate with automation. See discussion at https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/pull/114#issuecomment-1543385142

## Testing

These changes touch two code paths, `pod install` and the `gutenberg_dep_check` lane.

- CI verifies `pod install` works 
- Run `bundle exec fastlane gutenberg_dep_check` locally to verify it works. You should see:

<img width="1728" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/28a1ccf6-5f70-4e81-b0c6-967df835d726">


## Regression Notes

1. Potential unintended areas of impact – The automation on the Gutenbreg side will need to be update as well
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

**UI changes testing checklist:** Not a UI PR.

mokagio committed:
Rewrite Gutenberg configuration file from Ruby to YAML

This should prove easier to manipulate with automation.

mokagio committed:
Add note regarding `Gutenberg/config.yml` structure and automation

mokagio commented on Gutenberg/config.yml:
```diff
@@ -10,10 +8,12 @@
+  #
+  # We have automation that reads and updates this file.
+  # Leaving commented keys is discouraged, because as the automation would ignore them and the resulting updates would be noisy
+  tag: v1.101.0
```
Is this note going to create confusion down the track?

Alternatively, we could always specify both keys and:

1. Leave the unused one empty / `nil`
2. Maintain the `tag` before `commit` rule

Keen to hear what you think.

dcalhoun replied:
I think the noted approach should be fine. 👍🏻

crazytonyli commented on fastlane/lanes/release.rb:
```diff
@@ -228,20 +228,34 @@ def trigger_buildkite_release_build(branch:, beta:)
-    # If a tag is present, the commit value is ignored
-    UI.message("Gutenberg version: #{GUTENBERG_CONFIG[:tag]}")
+  rescue ArgumentError => e
+    UI.user_error!("Could not parse config YAML. Failed with: #{e.message}")
```
What you think reducing the scope of this rescue to just the YAML parsing `YAML.safe_load` part?

mokagio replied:
Good point. Addressed in a7500d0.

I tested it by changing the config to an invalid one (unterminated `""`)

```diff
 # Example:
 #
 #   LOCAL_GUTENBERG=../my-gutenberg-fork bundle exec pod install
-ref:
-  # This should have either a commit or tag key.
-  # If both are set, tag will take precedence.
-  #
-  # We have automation that reads and updates this file.
-  # Leaving commented keys is discouraged, because as the automation would ignore them and the resulting updates would be noisy
-  tag: v1.101.0
-github_org: wordpress-mobile
-repo_name: gutenberg-mobile
+#
+foo: "bar
```
and verified the error reporting:

<img width="1921" alt="Screenshot 2023-08-17 at 12 48 40 pm" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1218433/471fcfad-eda8-4a7b-b7e2-3da0f2000c15">

mokagio committed:
Only wrap `YAML.safe_load` in `begin - rescue`

Co-authored-by: Tony Li <tony.li@automattic.com>

See
https://github.com/wordpress-mobile/WordPress-iOS/pull/21191#discussion_r1296576723

dcalhoun reviewed (commented):
The proposed changes make sense to me. 👍🏻 It will be important to merge this in unison with the sibling https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/pull/142 to avoid release generation issues.

mokagio committed:
Merge 'origin/trunk' into mokagio/gutenberg-yaml-settings

mokagio committed:
Extract Gutenberg `config.yml` path to constant

This was just to silence RuboCop complaining about:

Assignment Branch Condition size for gutenberg_pod is too high. [<3, 17, 5> 17.97/17]
